[Routing] Clarify route aliases examples with explicit route names

welcoMattic opened the pull request:
Refers to https://github.com/symfony/symfony-docs/pull/20638#discussion_r1962144163

Actual examples for route aliasing are not crystal clear.

Now, examples in YAML, XML and PHP provide

- An original route, declared with an explicit name
- An alias to this route, with an explicit name

This change will allow to provide another clear example for route aliasing using Attribute in #20638 

cc @stof @damienfern

welcoMattic committed:
[Routing] Clarify route aliases examples with explicit route names

stof commented on routing.rst:
```diff
@@ -1419,9 +1434,12 @@ you decided not to maintain it anymore), you can deprecate its definition:
+        $routes->add('some_route_name', '/some_route_path')
+                ->controller('App\Controller\SomeController::index');
+        // second argument refers to the name of the route declared above
+        $routes->alias('alias_to_some_route_name', 'some_route_name');
```
```suggestion
        $routes->alias('alias_to_some_route_name', 'some_route_name')
```

welcoMattic commented:
@javiereguiluz this one could be merged to unlock https://github.com/symfony/symfony-docs/pull/20638

javiereguiluz commented:
I'm not sure about these changes. I think the previous description (old route + new route) was more clear.

The new names (some route + alias to some route) could be less clear, especially the new route name. Including "alias" as part of the route alias is unnecessarily long in my opinion and could make some readers think that the route alias must contain that `alias` prefix.

Let me ask @xabbuh and @OskarStark about this. Thanks!

welcoMattic commented:
> I'm not sure about these changes. I think the previous description (old route + new route) was more clear.
> 
> The new names (some route + alias to some route) could be less clear, especially the new route name. Including "alias" as part of the route alias is unnecessarily long in my opinion and could make some readers think that the route alias must contain that `alias` prefix.
> 
> Let me ask @xabbuh and @OskarStark about this. Thanks!

I can replace "alias" with "new", but it implies underlying use case: replacing the `some_route` usage with the alias. In my mind aliases are not necessarily here to replace existing routes, but to co-exists.
Imagine 2 websites sharing the same codebase except templates, 1 for BrandA and another for BrandB. Route aliases can be used to name route according to the Brand name in dedicated templates.

We can add a site note to mention the alias name is not mandatory to contain "alias" word.

These changes are on purpose to clarify the behavior using the Attribute way to alias route.

stof commented:
If you want to replace, you will generaly define the route with the new name and add a BC alias for the old name (especially when you want to deprecate it). That was the main confusion with the existing example. `new_route_name` is the deprecated name.

welcoMattic commented:
I think we do not have to presume what users will do with route aliasing feature (e.g creating new route and create an alias with the old route name). Just showing how to generally use route aliasing with generic terms, and then users will adapt to their needs/preferences. WDYT?

EDIT: I've update the PR by adding a small paragaph to explain clearly how route aliasing can be used to provide BC for renamed routes.

cc @javiereguiluz

javiereguiluz commented:
Maybe it's me, but this looks very confusing:

![](https://github.com/user-attachments/assets/df517ee5-d0e3-4ea5-86d5-7a4e8c767b18)

You create an alias ... and immediately deprecate it. The alias, being a new thing that you add, feels like the new route name, not the legacy route name.

-----

I still think we should use names like "new_route_name" and "original_route_name" or "legacy_route_name". It's true that this alias feature can be used in different ways, but I think we should focus on its main usage and use terms that are impossible to misunderstand.

damienfern commented:
I understand @javiereguiluz POV because I had the same feeling about aliases. Isn't this a sign that aliases on routes are plugged in a wrong way ?

IMHO, here is how it should be (a first draft, I don't know if it's doable): 

```yaml
some_route_name:
    path: /some-path
    controller: App\Controller\SomeController::index
    alias: 
          alias_to_some_route_name:
              # this outputs the following generic deprecation message:
              # Since acme/package 1.2: The "alias_to_some_route_name" route alias is deprecated. You should stop using it, as it will be removed in the future.
              deprecated:
                  package: 'acme/package'
                  version: '1.2'
```

Jean-Beru commented:
IMO, the [blog post](https://symfony.com/blog/new-in-symfony-5-4-route-aliasing#deprecating-routes) explains clearly the concept of deprecating a route. We could update the documentation in that way. eg:

My current route.
```yaml
some_route_name:
    path: /some-path
    controller: App\Controller\SomeController::index
```

I want to deprecate it in favor of `new_route_name`.
```yaml
some_route_name:
    alias: new_route_name
    # this outputs the following generic deprecation message:
    # Since acme/package 1.2: The "some_route_name" route alias is deprecated. You should stop using it, as it will be removed in the future.
    deprecated:
        package: 'acme/package'
        version: '1.2'

new_route_name:
    path: /some-path
    controller: App\Controller\SomeController::index
```

welcoMattic commented:
I've addressed comments, now examples are more explicit and I hope clearer:

First code block with initial route definition
```yaml
# config/routes.yaml
some_route_name:
    path: /some-path
    controller: App\Controller\SomeController::index
```

Second code block with an alias to the initial route

```yaml
# config/routes.yaml
some_route_name:
    path: /some-path
    controller: App\Controller\SomeController::index

new_route_name:
    # "alias" option refers to the name of the route declared above
    alias: some_route_name
```

Third code block where we "invert" the alias to be able to deprecate the initial route

```yaml
# Move the concrete route definition under ``new_route_name``
new_route_name:
    path: /some-path
    controller: App\Controller\SomeController::index

# Define the alias and the deprecation under the ``some_route_name`` definition
some_route_name:
    alias: new_route_name

    # this outputs the following generic deprecation message:
    # Since acme/package 1.2: The "some_route_name" route alias is deprecated. You should stop using it, as it will be removed in the future.
    deprecated:
        package: 'acme/package'
        version: '1.2'

    # or

    # you can also define a custom deprecation message (%alias_id% placeholder is available)
    deprecated:
        package: 'acme/package'
        version: '1.2'
        message: 'The "%alias_id%" route alias is deprecated. Do not use it anymore.'
```

I've also rephrased text paragraphs before and after each code block to explicitly explain what is the goal in each code block.

@javiereguiluz @Jean-Beru @damienfern @stof let me know if this version is good to you 🙏

Jean-Beru commented on routing.rst:
```diff
@@ -1432,8 +1508,8 @@ you decided not to maintain it anymore), you can deprecate its definition:
-In this example, every time the ``new_route_name`` alias is used, a deprecation
-warning is triggered, advising you to stop using that alias.
+In this example, every time the ``some_route_name`` alias is used, a deprecation
+warning is triggered, advising you to stop using this route and prefer using ``new_route_name``.
```
We could update the deprecation messages in examples to encourage using the new route.

Jean-Beru commented on routing.rst:
```diff
@@ -1364,38 +1409,58 @@ Route alias allow you to have multiple name for the same route:
+Now, let's say you want to replace the ``some_route_name`` route in favor of
+``new_route_name`` and mark the old one as deprecated.
+
+In the previous example, the alias was ``new_route_name`` was pointing to
```
```suggestion
In the previous example, the alias ``new_route_name`` was pointing to
```

OskarStark commented on routing.rst:
```diff
@@ -1338,13 +1338,56 @@ Route Aliasing
 
 Route alias allow you to have multiple name for the same route:
```
```suggestion
Route alias allow you to have multiple name for the same route.
```

OskarStark commented on routing.rst:
```diff
@@ -1338,13 +1338,56 @@ Route Aliasing
 
 Route alias allow you to have multiple name for the same route:
 
```
```suggestion
```

OskarStark commented:
Can we use something like `product_show` as old and `product_detail` as aliased route to add more real world feeling?

welcoMattic commented:
@OskarStark thanks for you review. I've fixed minor comments and renamed example routes.

I think we hit the most clear version we can:
- Real life route names
- Not confusing naming
- Clear explaination of each code block
- Common thread between Route Alias and Route Deprecation features

OskarStark commented on routing.rst:
```diff
@@ -1364,38 +1366,100 @@ Route alias allow you to have multiple name for the same route:
+Now, let's say you want to create a new route called ``product_details``
+that acts exactly the same as ``product_show``.
+
+Instead of duplicating the original route, you can create an alias for it. You can do this as follows:
```
```suggestion
Instead of duplicating the original route, you can create an alias for it:
```

OskarStark commented on routing.rst:
```diff
@@ -1364,38 +1366,100 @@ Route alias allow you to have multiple name for the same route:
+
+        # config/routes.yaml
+        product_show:
+            path: /some-path
```
```suggestion
            path: /product/{id}
```

OskarStark commented on routing.rst:
```diff
@@ -1364,38 +1366,100 @@ Route alias allow you to have multiple name for the same route:
 
+            # or
+
             # you can also define a custom deprecation message (%alias_id% placeholder is available)
```
```suggestion
            # you can define a custom deprecation message (%alias_id% placeholder is available)
```

OskarStark commented on routing.rst:
```diff
@@ -1405,35 +1469,46 @@ you decided not to maintain it anymore), you can deprecate its definition:
 
+            // or
+
             // you can also define a custom deprecation message (%alias_id% placeholder is available)
```
```suggestion
            // you can define a custom deprecation message (%alias_id% placeholder is available)
```

OskarStark commented on routing.rst:
```diff
@@ -1364,38 +1366,100 @@ Route alias allow you to have multiple name for the same route:
+                https://symfony.com/schema/routing/routing-1.0.xsd">
+
+            <route id="product_show" path="/product/{id}" controller="App\Controller\ProductController::show"/>
+            <!-- "alias" attribute value refers to the name of the route declared above -->
+            <route id="product_details" alias="product_show"/>
+        </routes>
+
+    .. code-block:: php
+
+        // config/routes.php
+        use Symfony\Component\Routing\Loader\Configurator\RoutingConfigurator;
+
+        return static function (RoutingConfigurator $routes): void {
+            $routes->add('product_show', '/product/{id}')
+                    ->controller('App\Controller\ProductController::show');
+            // second argument refers to the name of the route declared above
+            $routes->alias('product_details', 'product_show');
+        };
+
+In this example, both ``product_show`` and ``product_details`` routes can
 be used in the application and will produce the same result.
 
 .. _routing-alias-deprecation:
 
 Deprecating Route Aliases
 ~~~~~~~~~~~~~~~~~~~~~~~~~
 
-If some route alias should no longer be used (because it is outdated or
-you decided not to maintain it anymore), you can deprecate its definition:
+Route aliases can be used to provide backward compatibility for routes that
+have been renamed.
+
+Now, let's say you want to replace the ``product_show`` route in favor of
+``product_details`` and mark the old one as deprecated.
+
+In the previous example, the alias ``product_details`` was pointing to
+``product_show`` route.
+
+As you want to deprecate the ``product_show`` route, so let's invert the alias as follows
+to be able to mark it as deprecated using the ``deprecated`` option:
```
Besides the minor comments, this is the only part which has a lot of complexity and back in forth imho, can we shorten this somehow and more "straightforward"?
Thanks

welcoMattic replied:
I understand your point.
Here is a suggestion:

To mark the ``product_show`` route as deprecated, you need to "switch" the alias.
The ``product_show`` become the alias, and will now point to the ``product_details`` route.
This way, the ``product_show`` alias could be deprecated.

OskarStark replied:
Much better and shorter. Love it ❤️

damienfern commented on routing.rst:
```diff
@@ -1336,15 +1336,17 @@ A possible solution is to change the parameter requirements to be more permissiv
 --------------
 
-Route alias allow you to have multiple name for the same route:
+Route alias allow you to have multiple name for the same route.
```
```suggestion
Route alias allows you to have multiple name for the same route.
```

xabbuh replied:
Maybe instead:
 
```suggestion
Route aliases allow you to have multiple name for the same route.
```

damienfern commented on routing.rst:
```diff
@@ -1364,38 +1366,100 @@ Route alias allow you to have multiple name for the same route:
 
-If some route alias should no longer be used (because it is outdated or
-you decided not to maintain it anymore), you can deprecate its definition:
+Route aliases can be used to provide backward compatibility for routes that
+have been renamed.
```
IMHO, this part should be on top of the doc file as it is not specific to route deprecation, but it is related to route alias. 

An example : 
```diff
- Route alias allows you to have multiple name for the same route.
+ Route alias allows you to have multiple name for the same route 
+ and can be used to provide backward compatibility for routes that
+ have been renamed.
```

welcoMattic replied:
I agree to add precision about BC. For the position, it's already on top on Route Alias chapter.

welcoMattic commented:
Merging this one will unlock https://github.com/symfony/symfony-docs/pull/20638 and allow me to finish Route Alias documentation refreshing 🙏

OskarStark commented:
@symfony/docs can we get another approval please? Thanks

javiereguiluz commented on routing.rst:
```diff
@@ -1336,15 +1336,19 @@ A possible solution is to change the parameter requirements to be more permissiv
 --------------
 
-Route alias allow you to have multiple name for the same route:
+Route alias allows you to have multiple name for the same route
```
`multiple name` -> `multiple names`

javiereguiluz commented on routing.rst:
```diff
@@ -1336,15 +1336,19 @@ A possible solution is to change the parameter requirements to be more permissiv
+Route alias allows you to have multiple name for the same route
+and can be used to provide backward compatibility for routes that
+have been renamed.
+Let's say you have a route called ``product_show``
```
Let's add a `:` at the end of this phrase. Also, we can move this phrase to the previous line to be part of the first paragraph.

javiereguiluz reviewed (approved):
The new docs are now super clear to me! Thanks a lot Mathieu for the patience during the review process.

stof commented on routing.rst:
```diff
@@ -1336,15 +1336,19 @@ A possible solution is to change the parameter requirements to be more permissiv
-Route alias allow you to have multiple name for the same route:
+Route alias allows you to have multiple name for the same route
+and can be used to provide backward compatibility for routes that
+have been renamed.
```
should we mention that this BC is only for route generation ? If you inspect the `_route` request attribute after url matching, the canonical name of the route will always be used here, not the name of an alias.

welcoMattic replied:
👍 for me.

I think we could focus on this after complete update of Route Alias (including https://github.com/symfony/symfony-docs/pull/20638)?

We could address this topic individually on Route Alias and Deprecation paragraphs to indicate in each case what will be the value of `_route` requrest attribute. 

The current PR is more about clarification of the existing documentation.

welcoMattic commented:
@symfony/docs comments addressed, ready to be merged 🎉

OskarStark commented:
Thank you Mathieu.
